google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/internal/RowSetUtil.java

```diff
@@ -0,0 +1,346 @@
+    }
+    if (!isSegmentEmpty) {
+      results.add(segment.build());
+    } else if (allowEmptySplit) {
+      results.add(null);
+    }
+
+    return results;
+  }
+
+  /** Get the bounding range of a {@link RowSet}. */
+  public static ByteStringRange getBound(RowSet rowSet) {
+    // Find min & max keys
+    ByteString minKey = null;
+    ByteString maxKey = null;
+
+    for (ByteString key : rowSet.getRowKeysList()) {
+      if (minKey == null || ByteStringComparator.INSTANCE.compare(minKey, key) > 0) {
+        minKey = key;
+      }
+      if (maxKey == null || ByteStringComparator.INSTANCE.compare(maxKey, key) < 0) {
+        maxKey = key;
+      }
+    }
+
+    // Convert min & max keys in start & end points for a range
+    StartPoint minStartPoint = null;
+    EndPoint maxEndPoint = null;
+    if (minKey != null) {
+      minStartPoint = new StartPoint(minKey, true);
+    }
+    if (maxKey != null) {
+      maxEndPoint = new EndPoint(maxKey, true);
+    }
+
+    // Expand the range using the RowSet ranges
+    for (RowRange rowRange : rowSet.getRowRangesList()) {
+      StartPoint currentStartPoint = StartPoint.extract(rowRange);
+      if (minStartPoint == null || minStartPoint.compareTo(currentStartPoint) > 0) {
+        minStartPoint = currentStartPoint;
+      }
+
+      EndPoint currentEndpoint = EndPoint.extract(rowRange);
+      if (maxEndPoint == null || maxEndPoint.compareTo(currentEndpoint) < 0) {
+        maxEndPoint = currentEndpoint;
+      }
+    }
+
+    // Build a range using the endpoints
+    ByteStringRange boundingRange = ByteStringRange.unbounded();
+    if (minStartPoint != null) {
+      if (minStartPoint.isClosed) {
+        boundingRange.startClosed(minStartPoint.value);
+      } else {
+        boundingRange.startOpen(minStartPoint.value);
+      }
+    }
+    if (maxEndPoint != null) {
+      if (maxEndPoint.isClosed) {
+        boundingRange.endClosed(maxEndPoint.value);
+      } else {
+        boundingRange.endOpen(maxEndPoint.value);
+      }
+    }
+
+    return boundingRange;
+  }
+
+  private static final Comparator<RowRange> RANGE_START_COMPARATOR =
+      new Comparator<RowRange>() {
+        @Override
+        public int compare(@Nonnull RowRange o1, @Nonnull RowRange o2) {
+          return StartPoint.extract(o1).compareTo(StartPoint.extract(o2));
+        }
+      };
+
+  /** Helper class to ease comparison of RowRange start points. */
+  private static final class StartPoint implements Comparable<StartPoint> {
+    private final ByteString value;
+    private final boolean isClosed;
+
+    @Nonnull
+    static StartPoint extract(@Nonnull RowRange rowRange) {
+      switch (rowRange.getStartKeyCase()) {
+        case STARTKEY_NOT_SET:
+          return new StartPoint(ByteString.EMPTY, true);
+        case START_KEY_CLOSED:
+          return new StartPoint(rowRange.getStartKeyClosed(), true);
+        case START_KEY_OPEN:
+          // Take care to normalize an open empty start key to be closed.
```

Comment:
Very uneasy about this translation. 

For example, translating back will give us wrong start key.

if (minStartPoint.isClosed) {
       boundingRange.startClosed(minStartPoint.value);
 }

Reply:
I'm not sure what you mean here. I updated the code to be explicit about handling the empty string case. But I don't understand what you mean about translating back giving the wrong start key.